ethcore/src/client/client.rs

```diff
@@ -1853,21 +1853,96 @@ impl BlockChainClient for Client {
 	}
 
 	fn logs(&self, filter: Filter) -> Vec<LocalizedLogEntry> {
-		let (from, to) = match (self.block_number_ref(&filter.from_block), self.block_number_ref(&filter.to_block)) {
-			(Some(from), Some(to)) => (from, to),
-			_ => return Vec::new(),
-		};
+		macro_rules! return_empty_if_none {
+			( $v: expr ) => (
+				match $v {
+					Some(value) => value,
+					None => return Vec::new(),
+				}
+			)
+		}
 
 		let chain = self.chain.read();
-		let blocks = filter.bloom_possibilities().iter()
-			.map(move |bloom| {
-				chain.blocks_with_bloom(bloom, from, to)
-			})
-			.flat_map(|m| m)
-			// remove duplicate elements
-			.collect::<HashSet<u64>>()
-			.into_iter()
-			.collect::<Vec<u64>>();
+
+		// First, check whether `filter.from_block` and `filter.to_block` is on the canon chain. If so, we can use the
+		// optimized version.
+		let is_canon = |id| {
+			match id {
+				// If it is referred by number, then it is always on the canon chain.
+				&BlockId::Earliest | &BlockId::Latest | &BlockId::Number(_) => Some(true),
+				// If it is referred by hash, we see whether a hash -> number -> hash conversion gives us the same
+				// result.
+				&BlockId::Hash(hash) =>
+					Some(hash == chain.block_hash(chain.block_number(&hash)?)?),
+			}
+		};
+
+		let blocks = if return_empty_if_none!(is_canon(&filter.from_block)) && return_empty_if_none!(is_canon(&filter.to_block)) {
+			// If we are on the canon chain, use bloom filter to fetch required hashes.
+			let from = return_empty_if_none!(self.block_number_ref(&filter.from_block));
+			let to = return_empty_if_none!(self.block_number_ref(&filter.to_block));
+
+			let mut numbers = filter.bloom_possibilities().iter()
+				.map(|bloom| {
+					chain.blocks_with_bloom(bloom, from, to)
+				})
+				.flat_map(|m| m)
+				// remove duplicate elements
+				.collect::<HashSet<u64>>()
```

Comment:
I think instead of `HashSet` + `sort` we could collect to `BTreeSet<u64>` and then convert to `Vec<H256>`